iceoryx_examples/automotive_soa/include/owl/kom/field_subscriber.inl

```diff
@@ -0,0 +1,172 @@
+
+    while (m_subscriber.take()
+               .and_then([&](const auto& sample) {
+                   callable(sample.get());
+                   numberOfSamples++;
+               })
+               .or_else([](auto& result) {
+                   if (result != iox::popo::ChunkReceiveResult::NO_CHUNK_AVAILABLE)
+                   {
+                       std::cerr << "Error receiving chunk!" << std::endl;
+                   }
+               })
+           && numberOfSamples < maxNumberOfSamples)
+    {
+    }
+    return numberOfSamples;
+}
+
+template <typename T>
+inline Future<T> FieldSubscriber<T>::Get()
+{
+    // If we call Get() twice shortly after each other, once the response of the first request has not yet
+    // arrived, we have a problem
+    //! [FieldSubscriber get]
+    bool requestSuccessfullySent{false};
+    m_client.loan()
+        .and_then([&](auto& request) {
+            request.getRequestHeader().setSequenceId(m_sequenceId);
+            request.send().and_then([&]() { requestSuccessfullySent = true; }).or_else([](auto& error) {
+                std::cerr << "Could not send Request! Error: " << error << std::endl;
+            });
+        })
+        .or_else([&](auto& error) {
+            std::cerr << "Could not allocate Request! Error: " << error << std::endl;
+            requestSuccessfullySent = false;
+        });
+
+    if (!requestSuccessfullySent)
+    {
+        return Future<FieldType>();
+    }
+    return receiveResponse();
+    //! [FieldSubscriber get]
+}
+
+template <typename T>
+inline Future<T> FieldSubscriber<T>::Set(const FieldType& value)
+{
+    // If we call Set() twice shortly after each other, once the response of the first request has not yet
+    // arrived, we have a problem
+    bool requestSuccessfullySent{false};
+    m_client
+        .loan()
+        //! [FieldSubscriber set]
+        .and_then([&](auto& request) {
+            request.getRequestHeader().setSequenceId(m_sequenceId);
+            request->emplace(value);
+            //! [FieldSubscriber set]
+            request.send().and_then([&]() { requestSuccessfullySent = true; }).or_else([](auto& error) {
+                std::cerr << "Could not send Request! Error: " << error << std::endl;
+            });
+        })
+        .or_else([&](auto& error) {
+            std::cerr << "Could not allocate Request! Error: " << error << std::endl;
+            requestSuccessfullySent = false;
+        });
+
+    if (!requestSuccessfullySent)
+    {
+        return Future<FieldType>();
+    }
+    return receiveResponse();
+}
+
+template <typename T>
+inline Future<T> FieldSubscriber<T>::receiveResponse()
+{
+    Promise<FieldType> promise;
+    auto future = promise.get_future();
+    // Typically you would e.g. use a worker pool here, for simplicity we use a plain thread
+    std::thread(
+        [&](Promise<FieldType>&& promise) {
+            // Avoid race if MethodClient d'tor is called while this thread is still running
```

Comment:
@MatthiasKillat found a race condition here which is not solved by the guard (I am mentioning this here since the comment seems to not be listed.)

The problem is the following:
1. Someone calls `receiveResponse`. The thread is stalled by the OS in line 134
2. After the method returns the destructor of `FieldSubscriber<T>` is called.
3. The destructor locks successfully the mutex and destroys the object
4. Now the thread continues and locks an already destroyed mutex.

We discussed this and one idea to solve this evil race condition is to introduce an atomic counter. Before the thread is started we increment the counter by one and then create the thread. The last statement in the thread is the decrement of that counter.
In the destructor we create a busy loop where we loop until that counter has reached zero and all problems are solved.